compact_str/src/repr2/mod.rs

```diff
@@ -0,0 +1,856 @@
+use core::str::Utf8Error;
+use core::{
+    mem,
+    ptr,
+};
+use std::borrow::Cow;
+
+#[cfg(feature = "bytes")]
+mod bytes;
+
+mod capacity;
+mod heap;
+mod inline;
+mod iter;
+mod nonmax;
+mod num;
+mod traits;
+
+use capacity::Capacity;
+use heap::HeapBuffer;
+use inline::InlineBuffer;
+use nonmax::NonMaxU8;
+pub use traits::IntoRepr;
+
+/// The max size of a string we can fit inline
+pub const MAX_SIZE: usize = std::mem::size_of::<String>();
+/// Used as a discriminant to identify different variants
+pub const HEAP_MASK: u8 = 0b11111110;
+/// When our string is stored inline, we represent the length of the string in the last byte, offset
+/// by `LENGTH_MASK`
+pub const LENGTH_MASK: u8 = 0b11000000;
+
+const EMPTY: Repr = Repr::new_inline("");
+
+#[repr(C)]
+pub struct Repr(
+    // We have a pointer in the repesentation to properly carry provenance
+    *const (),
```

Comment:
I think it's possible to use `ptr::NonNull<u8>` here if we modify `push_str`, `extend` and etc to ignore null bytes inserted into the string.

Reply:
Good thought! For now I'll just keep it `*const ()` but we could change this in the future